src/macaron/vsa/vsa.py

```diff
@@ -0,0 +1,251 @@
+                "content": policy_content,
+            },
+            verificationResult=verification_result,
+            verifiedLevels=[],
+        ),
+    )
+
+
+def get_subject_verification_result(policy_result: dict) -> tuple[str, VerificationResult] | None:
+    """Get the PURL (string) and verification result of the single software component the policy applies to.
+
+    This is currently done by reading the facts of two relations:
+    ``component_violates_policy``, and ``component_satisfies_policy``
+    from the result of the policy engine.
+
+    We define two PURLs to be different if the two PURL strings are different.
+
+    The result of this function depends on the policy engine result:
+
+    - If there exist multiple different PURLs, this function returns ``None``.
+    - If there exist multiple occurrences of the same PURL and it is the only unique
+      PURL in the policy engine result, this function returns the latest occurrence,
+      which is the PURL that goes with the highest component ID, taking advantage of
+      component IDs being auto-incremented.
+    - If there is no PURL in the result, i.e. the policy applies to no software component
+      in the database, this function also returns ``None``.
+
+    Parameters
+    ----------
+    policy_result : dict
+        The result of the policy engine, including two relations:
+        ``component_violates_policy``, and ``component_satisfies_policy``.
+
+    Returns
+    -------
+    tuple[str, VerificationResult] | None
+        A pair of PURL and verification result of the only software component that
+        the policy applies to, or ``None`` according to the aforementioned conditions.
+    """
+    component_violates_policy_facts = policy_result.get("component_violates_policy", [])
+    component_satisfies_policy_facts = policy_result.get("component_satisfies_policy", [])
+
+    # key: PURL; value: result with the highest component id
+    component_results: dict[str, tuple[int, VerificationResult]] = {}
+
+    for component_id_string, purl, _ in component_violates_policy_facts:
+        component_id = int(component_id_string)
+        if purl not in component_results:
+            component_results[purl] = (component_id, VerificationResult.FAILED)
+        else:
+            current_component_id, _ = component_results[purl]
+            if component_id > current_component_id:
+                component_results[purl] = (component_id, VerificationResult.FAILED)
+    for component_id_string, purl, _ in component_satisfies_policy_facts:
+        component_id = int(component_id_string)
+        if purl not in component_results:
+            component_results[purl] = (component_id, VerificationResult.PASSED)
+        else:
+            current_component_id, _ = component_results[purl]
+            if component_id > current_component_id:
+                component_results[purl] = (component_id, VerificationResult.PASSED)
+
+    if len(component_results) != 1:
```

Comment:
It would be good to differentiate `len(component_results) == 0` vs `len(component_results) > 1`and add a useful info log for the latter.

Reply:
See 450608c15d80c7b731eccba3c177a53872983b5.